tools/autograd/derivatives.yaml

```diff
@@ -217,17 +217,11 @@
   self: other.cross(grad, dim)
   other: grad.cross(self, dim)
 
-- name: cumprod(Tensor self, int64_t dim)
-  self: cumprod_backward(grad, self, dim)
+- name: cumprod(Tensor self, int64_t dim, *, ScalarType? dtype)
+  self: cumprod_backward(grad.to(self.type()), self, dim)
```

Comment:
For all these functions:
- does this generate an additional cast if we trace it for the case where a dtype isn't passed?

- Also, can you use scalar_type() instead of type()?

Reply:
It doesn't do a cast if a `dtype` has not been passed to the original forward call:

https://github.com/pytorch/pytorch/blob/517c7c98610402e2746586c78987c64c28e024aa/aten/src/ATen/templates/TensorMethods.h#L12-L16

Reply:
Sorry, by generate an additional cast I meant "trace an additional cast".

Reply:
Sorry for the noob questions. Do you mean JIT tracing the backwards path? In what circumstance will this happen? I've failed to create a test scenario so far.

Though, I assume it won't be traced if the sample input doesn't execute the copy path, which it won't.
```
>>> torch.tensor(1.0, requires_grad=True).max().to(torch.double).grad_fn
<CopyBackwards object at 0x7f695dd3f898>
>>> torch.tensor(1.0, requires_grad=True).max().to(torch.float).grad_fn
<MaxBackward1 object at 0x7f695dd3f860>
```